test/model/recent_senders_test.dart

```diff
@@ -0,0 +1,201 @@
+      final user2 = eg.user(userId: 2);
+
+      final messages = [
+        eg.streamMessage(stream: stream1, topic: 'a', sender: user1, id: 300),
+        eg.streamMessage(stream: stream1, topic: 'b', sender: user1, id: 100),
+        eg.streamMessage(stream: stream1, topic: 'a', sender: user2, id: 200),
+      ];
+
+      model.handleMessages(messages);
+      checkMatchesMessages(messages, model);
+
+      model.handleDeleteMessageEvent(
+        DeleteMessageEvent(id: 1, messageIds: [300, 200],
+          messageType: MessageType.stream, streamId: stream1.streamId, topic: 'a'),
+        Map.fromEntries(messages.map((msg) => MapEntry(msg.id, msg))));
+
+      checkMatchesMessages(
+        [eg.streamMessage(stream: stream1, topic: 'b', sender: user1, id: 100)],
+        model);
+    });
+
+    test('latestMessageIdOfSenderInStream', () {
+      final model = RecentSenders();
+      final stream1 = eg.stream(streamId: 1);
+      final stream2 = eg.stream(streamId: 2);
+      final user1 = eg.user(userId: 1);
+      final user2 = eg.user(userId: 2);
+
+      final messages = [
+        eg.streamMessage(stream: stream1, sender: user1, id: 100),
+        eg.streamMessage(stream: stream1, sender: user1, id: 300),
+        eg.streamMessage(stream: stream1, sender: user1, id: 200),
+      ];
+
+      model.handleMessages(messages);
+
+      check(model)
+        ..latestMessageIdOfSenderInStream(
+            streamId: stream1.streamId, senderId: user1.userId).equals(300)
+        ..latestMessageIdOfSenderInStream(
+            streamId: stream1.streamId, senderId: user2.userId).equals(null)
+        ..latestMessageIdOfSenderInStream(
+            streamId: stream2.streamId, senderId: user1.userId).equals(null);
+    });
+
+    test('latestMessageIdOfSenderInTopic', () {
+      final model = RecentSenders();
+      final stream1 = eg.stream(streamId: 1);
+      final stream2 = eg.stream(streamId: 2);
+      final user1 = eg.user(userId: 1);
+      final user2 = eg.user(userId: 2);
+
+      final messages = [
+        eg.streamMessage(stream: stream1, topic: 'a', sender: user1, id: 300),
+        eg.streamMessage(stream: stream1, topic: 'a', sender: user1, id: 200),
+      ];
+
+      model.handleMessages(messages);
+
+      check(model)
+        ..latestMessageIdOfSenderInTopic(streamId: stream1.streamId,
+            topic: 'a', senderId: user1.userId).equals(300)
+        ..latestMessageIdOfSenderInTopic(streamId: stream1.streamId,
+            topic: 'a', senderId: user2.userId).equals(null)
+        ..latestMessageIdOfSenderInTopic(streamId: stream1.streamId,
+            topic: 'a', senderId: user2.userId).equals(null)
+        ..latestMessageIdOfSenderInTopic(streamId: stream1.streamId,
+            topic: 'b', senderId: user1.userId).equals(null)
+        ..latestMessageIdOfSenderInTopic(streamId: stream2.streamId,
+            topic: 'a', senderId: user1.userId).equals(null);
+    });
+
+    group('MessageIdTracker', () {
```

Comment:
Do these three tests cover anything that isn't already covered by the tests above? It looks like they don't — I think each of them is very similar to some test above and is basically just testing the same thing.

So let's write each of those tests just once. I think the contents of these tests are successfully using the approach I described at https://github.com/zulip/zulip-flutter/pull/692#discussion_r1664646094 — but then at that point they can just be treated as more tests of `RecentSenders` :slightly_smiling_face:, and they can also be deduplicated with other `RecentSenders` tests that do the same thing.

Relatedly, `MessageIdTracker` can be made private. Or `@visibleForTesting` if the tests have a need for it, but I think after my comment above on `checkMatchesMessages` they won't.